pact/brackets-review.pact

```diff
@@ -0,0 +1,504 @@
+    games-won:list)
+
+  (deftable users-table:{users})
+
+  (defcap BRACKET-ADMIN-BB (admin-key:string bracket-name:string)
+    (with-read bracket-betting-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap BRACKET-ADMIN-EB (admin-key:string bracket-name:string)
+    (with-read empty-bracket-table bracket-name { "admin":= admin-key-db }
+      (enforce (= admin-key admin-key-db) "you are not the bracket admin")
+    )
+  )
+
+  (defcap USER-CAN-TRANSFER (from:string amount:integer)
+    (with-read users-table from { "balance" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap EB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read empty-bracket-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap BB-CONTRACT-CAN-TRANSFER (bracket-name:string amount:integer)
+    (with-read bracket-betting-table bracket-name { "moneyPool" := from-bal }
+      (enforce (>= from-bal amount) "Insufficient Funds")
+    )
+  )
+
+  (defcap IS-REGISTERED-USER (user-key:string)
+    (let ((users (keys users-table)))
+          ;should be user-id
+        (enforce (contains user-key users) "you are not a registered user")
+    )
+  )
+
+  ;needs to be passed as arg to function (the guard)
+  ;user-key is both id and documents user from payload
+  ;usually its keyset in the arguments
+  (defun init-user (user-key:string username:string)
+    ;user id
+    (insert users-table user-key {
+      "username": username,
+      "balance": 100,
+      "user-guard": (read-keyset user-key),
+      "bb-games": [],
+      "bb-admins": [],
+      "eb-games": [],
+      "eb-admins": [],
+      "games-won": []
+    })
+  )
+
+  (defun get-user-info (user-key:string)
+    (with-capability (IS-REGISTERED-USER user-key)
+        (with-read users-table user-key {
+          "username":=username,
+          "bb-games":=bb-games,
+          "bb-admins":=bb-admins,
+          "eb-games":=eb-games,
+          "eb-admins":=eb-admins,
+          "balance":= balance,
+          "games-won":= games-won}
+          [username, bb-games, bb-admins, eb-games, eb-admins, balance, games-won]
+        )
+    )
+  )
+
+  (defun get-all-users ()
+    ;;anyone can call this
+    (keys users-table)
+  )
+
+  (defun init-empty-bracket
+    (bracket-name:string bracket:list number-players:integer entry-fee:integer)
+    "initiate a new bracket"
+    ;sender is id not key...
+    (let ((admin-key (at "sender" (chain-data))))
+    ;is already a user of our systems
+    (with-capability (IS-REGISTERED-USER admin-key)
+      (enforce (check-bracket-validity bracket) "bracket format not valid")
+      ; anyone can init a new bracket.
+       (insert empty-bracket-table bracket-name {
+        ;"admin": (at "sender" (chain-data)),
+         "admin": admin-key,;should be a guard
```

Comment:
Agreed should be a guard, and this is easily done: just specify the keyset in the payload like any other transaction. 

The confusion might be that you're relying on `sender` a la Ethereum, which I wouldn't recommend for reasons:
- It means this can only run on a public, gas-paying blockchain. A private blockchain could function equally well, and doesn't have the PublicMeta payload.
- In general, it's more opaque and less flexible -- maybe in the future someone else would pay the gas and therefore be the `sender` but the bracket admin might be a different key.
- Even if the sender is always the bracket admin, it's still good practice to separately specify.

Thus, `init-empty-bracket` would:
1. Take the value for "admin" as an argument, not read from "sender". I would probably call this "admin-id" as an arg and in the db.
2. Read a keyset from the payload and insert it into the table. I would call this "admin-guard".

The pattern in pact is to always pair an id with a guard. It seems verbose but it ensures you're multi-sig and multi-guard ready.

Reply:
A more minimal/tactical change would be to keep using "sender" for now, but add the guard to the payload and read it as specified above.